cxx/util_io.cc

```diff
@@ -31,6 +31,34 @@ void verify_noisy_relation_domains(const T_schema& schema) {
   }
 }
 
+#define TYPE_CHECK(i, expected_type) \
```

Comment:
Why use macros instead of functions here?

Reply:
So, partly out of habit -- before C++ had closures, macros were a cheap way of not having to list out all of the captured variables.

But macros have a bunch of other handy features that make them good for this sort of error reporting.  You often want to report the line number or function of the code the error happens at for example, and that's easily done within a macros with __LINE__.

Another feature, which I actually use here, is the fact that the input to a macro is treated as text and doesn't have to be a value.  So when I write "TYPE_CHECK(i, identifier)" below, identifier isn't a value, it's just a piece of text.  It becomes a value when TokenType:: is prepended to it, but it is also used textually inside the error reporting string.